src/icp_canister_backend/tests/test.rs

```diff
@@ -214,3 +215,179 @@ fn test_deposit_fails_without_transfer() {
         result
     );
 }
+
+#[test]
+fn test_successful_withdrawal() {
+    let (pic, canister_id, ledger_id) = setup();
+    let user = Principal::from_text("xkbqi-2qaaa-aaaah-qbpqq-cai").unwrap();
+    let timelock: u64 = 1; 
+    let deposit_amount = Nat::from(100_000_000u64);
+
+    //User calls get_deposit_subaccount
+    let subaccount_result = pic.query_call(
+        canister_id, 
+        user, 
+        "get_deposit_subaccount", 
+        encode_args((user, timelock)).unwrap(),
+    ).expect("Failed to get deposit subaccount");
+    let subaccount: [u8; 32] = decode_one(&subaccount_result).unwrap();
+    
+    //User transfers tokens to the subaccount
+    let transfer_args = TransferArg {
+        from_subaccount: None,
+        to: Account {
+            owner: canister_id,
+            subaccount: Some(subaccount.to_vec()),
+        },
+        amount: deposit_amount.clone(),
+        fee: Some(TRANSFER_FEE.clone()),
+        memo: None,
+        created_at_time: None,
+    };
+    let transfer_result = pic.update_call(
+        ledger_id,
+        user,
+        "icrc1_transfer",
+        encode_args((transfer_args,)).unwrap(),
+    ).expect("Failed to transfer tokens");
+    let transfer_result: TransferResult = decode_one(&transfer_result).unwrap();
+    assert!(matches!(transfer_result, TransferResult::Ok(_)), "Transfer failed: {:?}", transfer_result); 
+    //User calls deposit function
+    let deposit_result = pic.update_call(
+        canister_id,
+        user,
+        "deposit",
+        encode_args((user, timelock)).unwrap(),
+    ).expect("Failed to call deposit");
+    let result: Result<(), DepositError> = decode_one(&deposit_result).unwrap();
+    assert!(result.is_ok(), "Deposit failed: {:?}", result);
+    //  Advance time so timelock expires
+    pic.advance_time(std::time::Duration::from_secs(2));
+    // Now withdraw
+    let withdraw_result = pic.update_call(
+        canister_id,
+        user,
+        "withdraw",
+        encode_args((0u64,)).unwrap(),
+    ).expect("Failed to call withdraw");
+    let result: Result<Nat, WithdrawError> = decode_one(&withdraw_result).unwrap();
+    assert!(matches!(result, Ok(_)), "Withdraw failed: {:?}", result);
+}
+
+#[test]
+fn test_withdraw_invalid_principal() {
+    let (pic, canister_id, ledger_id) = setup();
+    let user = Principal::from_text("xkbqi-2qaaa-aaaah-qbpqq-cai").unwrap();
+    let other = Principal::from_text("aaaaa-aa").unwrap();
+    let timelock: u64 = 1;
+    let deposit_amount = Nat::from(100_000_000u64);
```

Comment:
Create a utils.rs file and a function there - create_deposit(pic, canister_id, user) - to remove all the code repetition for deposit creation, also move transfer_fee constant there.